numba/unicode.py

```diff
@@ -377,11 +377,18 @@ def _find(substr, s):
 
 @njit(_nrt=False)
 def _count(substr, s):
-    # Naive, slow string matching for now
     count = 0
-    for i in range(len(s) - len(substr) + 1):
+    i = 0
+    incr = len(substr)
```

Comment:
Does saving `len(substr)` in a variable improve performance ?

Reply:
@oleksandr-pavlyk, I'm understanding your question as, if we were to generally save `len(substr)` in a variable (e.g. `substrlen`) at the beginning of the function would we improve performance of the whole function? If that's not right please correct me.

In the context of a jit wrapped function the `len()` call is just an access to the `_len`. So if we saved the length of the substring in a local variable then we could save us the overhead of a function call and a dictionary look up at the expensive of using a little extra memory. Since the user is most likely using the jit decorator to gain maximum performance, a little extra memory is most likely a trade off we are willing to make.

If you agree, I will go ahead and make that change

Reply:
Yes, I had in mind saving the result in a variable, and reusing it. I think it's worth checking how much of a performance impact this change actually has, if only to gain an insight into how much numba is able to optimize on its own.

Reply:
Got it, agreed. I will put together some comparisons with `timeit` and report back

Reply:
I put together this experiment,
```python
import numba
import random
import string
import timeit


@numba.njit
def compiledcount(a, b):
    return a.count(b)


PATTERNS = (
    "aaaaaa",
    "aa"
    "abcdef",
    "cde",
)
search_string = ''.join(random.choice(PATTERNS) for _ in range(1000))
target = random.choice(PATTERNS)
NUM = 1000000


comptime = timeit.timeit(
    'compiledcount(search_string, target)',
    globals=globals(),
    number=NUM
)
print(comptime, "seconds to perform", NUM, "loops of compiled count")
print(comptime/NUM, "seconds per loop")
```
Here is the implementation of count that does not save off len(substr) into a variable
```python
@njit(_nrt=False)
def _count(substr, s):
    count = 0
    i = 0
    incr = len(substr)
    if not incr:
        incr = 1
    # Naive, slow string matching for now
    while i < len(s) - len(substr) + 1:
        if _cmp_region(s, i, substr, 0, len(substr)) == 0:
            count += 1
            i += incr
        else:
            i += 1
    return count
```
Using this implementation to run the script above,
```bash
(numbaenv) JB-rhel ~/personal/repos/numba (feature/unicode-type-count) $ python count_times.py
35.33945925699663 seconds to perform 1000000 loops of compiled count
3.533945925699663e-05 seconds per loop
```
Here is the count implementation _with_ saving len(substr) into a local variable,
```python
@njit(_nrt=False)
def _count(substr, s):
    count = 0
    i = 0
    substrlen = len(substr)
    incr = substrlen
    if not incr:
        incr = 1
    # Naive, slow string matching for now
    while i < len(s) - substrlen + 1:
        if _cmp_region(s, i, substr, 0, substrlen) == 0:
            count += 1
            i += incr
        else:
            i += 1
    return count
```
Using this implementation,
```bash
(numbaenv) JB-rhel ~/personal/repos/numba (feature/unicode-type-count) $ python count_times.py
37.131292551988736 seconds to perform 1000000 loops of compiled count
3.713129255198874e-05 seconds per loop
```
So, you can see that, for this particular experiment, not saving len(substr) into a local variable was actually slightly faster. At this point I'm not sure if this would hold for all workloads.

As a point of reference I tried running the same script but with the njit decorator removed,
```bash
(numbaenv) root@JB-rhel ~/personal/repos/numba (feature/unicode-type-count) $ python count_times.py
3.477520063999691 seconds to perform 1000000 loops of python count
3.477520063999691e-06 seconds per loop
```
Interestingly, removing the njit decorator and running the pure python count implementation gives us an order of magnitude speed up. Either this experiment is not demonstrating a good numba use case or the count implementation just needs to be sped up. When I have a bit of time I will go into CPython and see what their string matching looks like.

Reply:
[As expected the string matching in CPython is far fro naive (click me)](https://github.com/python/cpython/blob/e42b705188271da108de42b55d9344642170aa2b/Objects/stringlib/fastsearch.h)

I reran the experiment using a search string that is ~10 characters long rather than thousands of characters. Relative performance of CPython versus numba jit compiled is still an order of magnitude difference in favor of CPython.
```bash
(numbaenv) JB-rhel ~/personal/repos/numba (feature/unicode-type-count) $ python count_times.py
0.1575793359952513 seconds to perform 1000000 loops of python count
1.575793359952513e-07 seconds per loop
3.9455732989881653 seconds to perform 1000000 loops of compiled count
3.945573298988165e-06 seconds per loop
```